aten/src/ATen/native/quantized/cpu/kernels/QuantizedOpKernels.cpp

```diff
@@ -218,7 +218,187 @@ void qmaxpool_2d_nhwc_kernel(const Tensor &qx,
       }  // for row
     }  // for b
   });
-} 
+}
+
+void qadaptive_avg_pool2d_nhwc_kernel(const Tensor &qx,
+                                      Tensor &qy,
+                                      int64_t b,
+                                      int64_t sizeD,
+                                      int64_t isizeH,
+                                      int64_t isizeW,
+                                      int64_t osizeH,
+                                      int64_t osizeW,
+                                      int64_t istrideB,
+                                      int64_t istrideD,
+                                      int64_t istrideH,
+                                      int64_t istrideW) {
+  AT_DISPATCH_QINT_TYPES(qx.scalar_type(), "adaptive_avg_pool2d_nhwc", [&]() {
+    scalar_t *idata = static_cast<scalar_t*>(qx.data_ptr());
+    scalar_t *odata = static_cast<scalar_t*>(qy.data_ptr());
+    auto minimum = std::numeric_limits<scalar_t::underlying>::lowest();
+    auto maximum = std::numeric_limits<scalar_t::underlying>::max();
+    auto *i_p = reinterpret_cast<typename scalar_t::underlying*>(idata + b * istrideB);
+    for (int64_t oh = 0; oh < osizeH; oh++) {
+      int istartH = (int)std::floor((float)(oh * isizeH) / osizeH);
+      int iendH = (int)std::ceil((float)((oh + 1) * isizeH) / osizeH);
+      int kH = iendH - istartH;
+      float kHr = 1.0 / kH;
+      for (int64_t ow = 0; ow < osizeW; ow++) {
+        auto *o_p = reinterpret_cast<typename scalar_t::underlying*>(odata + b * osizeH * osizeW * sizeD + (oh * osizeW + ow) * sizeD);
+        int istartW = (int)std::floor((float)(ow * isizeW) / osizeW);
+        int iendW = (int)std::ceil((float)((ow + 1) * isizeW) / osizeW);;
+        int kW = iendW - istartW;
+        float kHWr = kHr / kW;
+        int size = kH * kW;
+        float multiplier = qx.q_scale() / qy.q_scale() / size;
+
+        int64_t c = 0;
+        // For int8 or uint8quantization, we implicitly use int32 as accumulation
+        // Or else, it will go to the slow path
+        // TODO: support 16bit, 32bit, and etc.
+        constexpr auto vec_width = Vec256<scalar_t>::size() / 4;
+        auto* internal_i_p = i_p + istartH * istrideH + istartW * istrideW;
+
+        // TODO: more vectorization with loop interleaving
+#ifdef __AVX2__
+        for (; c + vec_width <= sizeD && vec_width == 8; c += vec_width) {
```

Comment:
Pull `vec_width == 8` into a separate `if` statement around this loop. It doesn't make sense to have it in the loop condition